### PR TITLE
✨ Add global support for amp-recaptcha-input component.

### PR DESCRIPTION
**Pull request by krdwan**
Fixes: https://github.com/ampproject/amphtml/issues/26114

- Add a new optional global attribute which changes the endpoint from which a recaptcha script is requested
- The original endpoint is not available in all locales.
- Lives in amp-recaptcha-input component.  Tests and documentation included as well.

### CODE REVIEW

File: extensions/amp-recaptcha-input/amp-recaptcha-input.md [hunk 1]
```diff
@@ -78,6 +78,10 @@ grecaptcha.execute('reCAPTCHA_site_key', {action: 'reCAPTCHA_example_action'});
     <td width="40%"><strong>data-action (required)</strong></td>
     <td><a href="https://developers.google.com/recaptcha/docs/v3">reCAPTCHA v3</a> action to be executed on form submission.</td>
   </tr>
+  <tr>
+    <td width="40%"><strong>global [optional]</strong></td>
```
Comment (caroqliu) [inline extensions/amp-recaptcha-input/amp-recaptcha-input.md]: It looks like the other recaptcha-forwarded attributes are prefixed with `data-`, let's use this so the API is self-consistent.

File: extensions/amp-recaptcha-input/amp-recaptcha-input.md [hunk 2]
```diff
@@ -78,6 +78,10 @@ grecaptcha.execute('reCAPTCHA_site_key', {action: 'reCAPTCHA_example_action'});
...
+    <td>Include this optional attribute to change the endpoint from which the iframe loads the reCAPTCHA api script.  More info can be found here: <a href="https://developers.google.com/recaptcha/docs/faq#can-i-use-recaptcha-globally">reCAPTCHA FAQ</a>.</td>
```
Comment (caroqliu) [inline extensions/amp-recaptcha-input/amp-recaptcha-input.md]: From a documentation standpoint, when an attribute is optional it's important to describe what it does when it is present, and what it does by default:

`By default, the iframe loads the recaptcha api script using the 'www.google.com' endpoint. There are some situations when this is not accessible. When the "data-global" attribute is included, the component will load the script from the "www.recaptcha.net" endpoint instead. More info ... `

File: extensions/amp-recaptcha-input/0.1/test/test-amp-recaptcha-service.js [hunk 3]
```diff
@@ -122,6 +122,24 @@ describes.realWin(
       }
     );
 
+    it(
+      'should not allow elements to register,' +
+        ' if they do not all or none pass in the global attribute',
+      () => {
```
[the remark this answers is not in the document]
Comment (krdwan) [reply extensions/amp-recaptcha-input/0.1/test/test-amp-recaptcha-service.js:131]: Thanks for calling this out, async / await certainly does look clean.  I went with the promises style this time since the existing tests were all done this way so wanted to be consistent.

File: extensions/amp-recaptcha-input/0.1/test/test-amp-recaptcha-service.js [hunk 4]
```diff
@@ -122,6 +122,24 @@ describes.realWin(
...
+        expect(recaptchaService.registeredElementCount_).to.be.equal(0);
+        return recaptchaService.register(fakeSitekey, true).then(() => {
+          expect(recaptchaService.registeredElementCount_).to.be.equal(1);
+          expect(recaptchaService.iframe_).to.be.ok;
+
+          return recaptchaService.register(fakeSitekey, false).catch(err => {
```
[the remark this answers is not in the document]
Comment (krdwan) [reply extensions/amp-recaptcha-input/0.1/test/test-amp-recaptcha-service.js]: Thanks for the callout, prettify forced me to change it slightly from your above suggestion in terms of formatting.  I did chain the catch block after the then though so there's less nesting than before.  Thanks for recommending that.

File: extensions/amp-recaptcha-input/0.1/test/test-amp-recaptcha-service.js [hunk 5]
```diff
@@ -122,6 +122,24 @@ describes.realWin(
...
+            expect(err).to.be.ok;
```
[the remark this answers is not in the document]
Comment (krdwan) [reply extensions/amp-recaptcha-input/0.1/test/test-amp-recaptcha-service.js:141]: Done!  Added to this and the previous tests which both receive errors from the same method.

File: extensions/amp-recaptcha-input/amp-recaptcha-input.md (hunk 1 above)
Comment (krdwan) [reply extensions/amp-recaptcha-input/amp-recaptcha-input.md]: Updated to use `data-`, thanks!

File: extensions/amp-recaptcha-input/amp-recaptcha-input.md (hunk 2 above)
Comment (krdwan) [reply extensions/amp-recaptcha-input/amp-recaptcha-input.md]: Your example documentation is much more descriptive.  I've included this version with some minor edits!

Comment (krdwan) [reply extensions/amp-recaptcha-input/amp-recaptcha-input.md]: Your example documentation is much more descriptive.  I've included this version with some minor edits!

**Comment by krdwan** [comment]
> Looks good, included a few comments. 👍
> 
> A note about adding new attributes to an AMP component:
> While the component internals can always be changed to read an attribute, people can't actually use these attributes immediately on their valid AMP documents. For example, if you copied an example usage of `amp-recaptcha-input` with the `global` attribute into the [AMP validator](https://validator.ampproject.org/), you would see that it's currently disallowed.
> 
> In order to allow the attribute, we would need to change the [validator rules ](https://github.com/ampproject/amphtml/blob/master/extensions/amp-recaptcha-input/validator-amp-recaptcha-input.protoascii) for the component to accept `global` in its attribute list. Normally, for a change like this, the code reviewer would ask you to update the component `protoascii` as well as its associated [validator- test cases.](https://github.com/ampproject/amphtml/tree/master/extensions/amp-recaptcha-input/0.1/test)
> 
> The one caveat to this is that all attributes prefixed with `data-` are already marked valid in AMP. This is because they don't implicitly do anything in the [HTML spec](https://developer.mozilla.org/en-US/docs/Learn/HTML/Howto/Use_data_attributes). I included a comment in the documentation about how we should rename the `global` attribute to `data-global` to be self-consistent with its other exposed attributes that are passed along to the recaptcha service (`data-sitekey` and `data-action`). In that case we wouldn't need to make the changes to the validator since we get them for free. (If your wondering why the others are included anyway, it is because they specify `mandatory: true` which isn't the case for `data-global`.)
> 
> tldr: No validator changes needed once the component reads the `"data-global"` attribute instead of `"global"`.

Thanks for the very thorough review @caroqliu.  The detailed comments and context are extremely helpful for me to pick up on how things are done at AMP.  Thanks for taking the time to write everything out.  Please see me latest revision.  Thanks!

File: 3p/recaptcha.js [hunk 6]
```diff
@@ -99,6 +103,14 @@ export function initRecaptcha(recaptchaApiBaseUrl = RECAPTCHA_API_URL) {
     'The sitekey is required for the <amp-recaptcha-input> iframe'
   );
   sitekey = dataObject.sitekey;
+  // Determine the recaptcha api URL based on global attribute
```
[the remark this answers is not in the document]
Comment (krdwan) [reply 3p/recaptcha.js]: Thanks, updated!

File: 3p/recaptcha.js [hunk 7]
```diff
@@ -99,6 +103,14 @@ export function initRecaptcha(recaptchaApiBaseUrl = RECAPTCHA_API_URL) {
...
+  devAssert(
+    hasOwn(dataObject, 'global'),
+    'The global attribute is required for the <amp-recaptcha-input> iframe'
```
[the remark this answers is not in the document]
Comment (krdwan) [reply 3p/recaptcha.js]: Updated!
